File: src/linear_regression.py
import numpy as np

class LinearRegression:
    """
    Linear Regression from scratch using gradient descent.
    Supports multivariate regression:
        y = Xw + b
    """
    def __init__(self, lr: float, iterations: int) -> None:
        self.lr = lr
        self.iterations = iterations
        self.w = None
        self.b = None
        self.loss_logs = []
    
    def predict(self, X: np.ndarray) -> np.ndarray:
        # X.shape = (n_samples, n_features) 
        return X @ self.w + self.b
    
    def loss_function(self, y: np.ndarray, y_pred: np.ndarray) -> float:
        """
        Mean squared error loss function
        """
        n = len(y)
        return (1/n) * np.sum((y - y_pred) ** 2)
    
    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        """
        Model training using gradient descent to update the weights.
        w = w - lr*dw
        b = b - lr*db
        """
        n_samples, n_features = X.shape
        
        self.w = np.zeros(n_features)
        self.b = 0

        self.loss_logs = [] # Track losses

        for i in range(self.iterations):
            # y predictions from predict function
            y_pred = self.predict(X)

            loss = self.loss_function(y, y_pred) # calculate loss
            self.loss_logs.append(loss) # append losses to vector

            # gradients
            dw = (-2 / n_samples) * X.T @ (y - y_pred)  # shape: (n_features,)
            db = (-2 / n_samples) * np.sum(y - y_pred)

            # update weights
            self.w -= self.lr * dw
            self.b -= self.lr * db

            # compute loss and track loss
            if i % 1000 == 0:
                print(f"Iteration {i}: Loss = {loss}")
    
    def score(self, X: np.ndarray, y: np.ndarray) -> float:
        y_pred = self.predict(X)

        ssr = np.sum((y - y_pred)**2)
        ssto = np.sum((y - np.mean(y))**2)
        return 1 - (ssr/ssto) 

File: R/linear-regression.R
  ########### Linear Regression ###########

  #random lr dataset#
  data = read.csv("../Linear-Regression/data/test.csv")
  head(data)

  par(mfrow=c(1,2))
  plot(data$x, data$y)

  ###### Linear Regression using lm function ######
  model = lm(data$y ~ data$x)

  summary(model)
  #Adjusted R squared 0.9891

  # y = b0 + b1x
  b0 = model$coefficients[1]
  print(b0)

  b1 = model$coefficients[2]
  print(b1)

  #plot line usign abline
  plot(x=data$x, y=data$y)
  abline(model, col='red', lwd=2) 

  #plot using y=b0+b1*x
  plot(x=data$x, y=data$y)

  xhat = seq(min(data$x), max(data$x), length=1000)
  yhat = b0 + b1*xhat

  lines(xhat, yhat, col="deepskyblue", lwd=2)

  ###### Linear Regression from scratch using Gradient Descent ######
  ###y = m * x + b###
  x=data$x
  y=data$y

  m=0 
  b=0

  rate=0.0001 #learning rate
  iter=100000 #iterations

  #Gradient Descent Algorithm
  gd = function(x, y, m, b, rate, iter){
    n=length(x)
    losses=numeric(iter)
    
    #loop
    for (i in 1:iter){
      y_predicted = m*x + b
      
      #Gradients / Derivatives 
      gm = -(2/n)*sum(x*(y - y_predicted)) #derivative of MSE dL/dm
      gb = -(2/n)*sum(y - y_predicted) #derivative of MSE dL/db
      
      #new m & b parameters
      m=m - rate*gm
      b=b - rate*gb
      
      #MSE
      loss = (1/n) * sum((y - y_predicted)^2)
      losses[i] = loss
      #loss per 100 iterations completed
      if (i %% 1000 ==0){
        cat("Iteration:", i, "Cost:", loss, "\n")
      }
    }
    
    return(list(m=m, b=b, losses=losses))
    
  }

  result = gd(x, y, m, b, rate, iter)
  print(result)

  #y = m*x + b
  slope = result$m
  intercept = result$b

  #Compare results
  comparison = data.frame(Coefficient = c("Intercept", "Slope"),
    Gradient_Descent = c(intercept, slope),
    LM = c(b0, b1))

  print(comparison)

  #Compare results using plots
  plot(x, y, pch=20, main="Linear Regression: lm() function vs Gradient Descent")

  xhat = seq(min(x), max(x), length.out = 1000)

  #LM line
  lines(xhat, b0+b1*xhat, col="red", lwd=2)

  #Gradient Descent line
  lines(xhat, intercept+slope*xhat, col="blue", lwd=2, lty=2)

  #Legend
  legend("topleft", legend=c("lm() function", "Gradient Descent"), col=c("red", "blue"), lty=c(1,2), lwd=2)

  par(mfrow=c(1,2))
  #lm() function Plot
  plot(x, y, pch=20, main="Linear Regression: lm() function")
  abline(model, col='red', lwd=2) 

  #gradient descent plot
  plot(x, y, pch=20, main="Linear Regression: Gradient Descent")
  lines(xhat, intercept+slope*xhat, col="blue", lwd=2)
